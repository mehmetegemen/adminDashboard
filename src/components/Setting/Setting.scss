@import "../../master.scss";

.setting {
  display: flex;
  border-bottom: 1px solid $lightGray;
  margin-top:10px;
  padding: 20px;
  &__content {
    width: calc(100% - 100px);
    margin: auto 0;
    padding-right: 10%;
  }
  &__switch {
    width: 100px;
    margin-bottom: auto;
  }
}