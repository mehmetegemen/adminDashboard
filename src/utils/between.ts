export default (x: number, min: number, max: number) => x >= min && x <= max;
