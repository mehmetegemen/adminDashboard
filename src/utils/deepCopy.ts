// Very fast
export default (obj: any) => JSON.parse(JSON.stringify(obj));
