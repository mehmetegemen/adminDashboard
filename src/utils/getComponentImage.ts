import connect from "connect";
import finalhandler from "finalhandler";
import http from "http";
import { AddressInfo } from "net";
import puppeteer from "puppeteer";

const createServer = async (html: string) => {
  const app = connect();
  app.use(
    (request: any, response: any, next: connect.NextFunction) =>
      request.url === "/" ? response.end(html) : next()
  );
  app.use(finalhandler);

  const server = http.createServer(app);

  await new Promise((resolve, reject) => {
    const startServer = () => {
      server.once("error", (e: NodeJS.ErrnoException) => {
        if (e.code === "EADDRINUSE") {
          server.close(startServer);
        }
      });
      server.listen(0, (err?) => (err ? reject(err) : resolve()));
    };
    startServer();
  });

  return server;
};

const takeScreenshot = async (url: string) => {
  const browser = await puppeteer
    .launch(
      {
        args: ["--disable-lcd-text"],
        defaultViewport: { width: 1920, height: 1080 },
      },
    );
  const page = await browser.newPage();

  await page.goto(url, { waitUntil: "load" });

  const image = await page.screenshot();
  browser.close();

  return image;
};

const generateImage = async () => {
  // get HTML from JSDOM
  const html = document.documentElement.outerHTML;

  const server = await createServer(html);
  // We created intermediary variables to make necessary Typescript assertions
  const address = server.address()! as AddressInfo;
  const port = address.port;
  const url = `http://localhost:${port}`;
  const screenshot = await takeScreenshot(url);
  await new Promise((resolve) => server.close(resolve));
  return screenshot;
};

export default generateImage;
