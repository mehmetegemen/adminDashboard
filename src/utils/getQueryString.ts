  export default ({ query, search }: { query: string, search: string}) => {
    // E.g. /(?:search\=)([^\&]*)/; gets ?search=something&...
    const searchQueryRegexp = new RegExp("(?:" + query + "\=)([^\&]*)");
    const searchQuery = searchQueryRegexp.exec(search);
    return searchQuery ? searchQuery[1] : "";
  };
