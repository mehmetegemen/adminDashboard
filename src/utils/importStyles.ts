import fs from "fs";
import sass from "node-sass";

interface IGenerateStyleProp {
  filePaths: string[];
  includePaths: string[];
}

export const generateStyle = async (props: IGenerateStyleProp) => {
  const { filePaths, includePaths } = props;
  let css = "";
  for (const filePath of filePaths) {
    const cssText = await new Promise<string>((resolve, reject) => {
      fs.readFile(filePath, (err, data) => {
        if (err) {
          return reject(err);
        }
        resolve(data.toString());
      });
    });
    const processedSassObject = await new Promise<sass.Result>((resolve, reject) => {
      sass.render(
        {
          data: cssText,
          includePaths: [...includePaths],
        },
        (err, result) => {
          if (err) {
            return reject(err);
          }
          resolve(result);
        },
      );
    });
    const processedCss = processedSassObject.css.toString();
    css += `\n${processedCss}`;
  }

  const head = document.getElementsByTagName("head")[0];
  const style = document.createElement("style");
  style.type = "text/css";
  style.innerHTML = css;
  head.appendChild(style);
};

export const removeStyle = () => {
  const styleElem = document.getElementsByTagName("style")[0];
  styleElem.parentNode!.removeChild(styleElem);
};
