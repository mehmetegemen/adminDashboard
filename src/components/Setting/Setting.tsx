import React, { FunctionComponent } from "react";
import SwitchButton from "../SwitchButton/SwitchButton";

import "./Setting.scss";

const Setting: FunctionComponent<{
  active: boolean,
}> = (props) => {
  const { active, children } = props;
  return (
    <div className="setting">
      <div className="setting__content">
        {children}
      </div>
      <div className="setting__switch">
        <SwitchButton 
          isChecked={active}
        />
      </div>
    </div>
  )
}

export default Setting;
