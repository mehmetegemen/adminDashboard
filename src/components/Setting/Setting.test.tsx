import { mount, shallow } from "enzyme";
import React from "react";
import ReactDOM from "react-dom";
import getComponentImage from "../../utils/getComponentImage";
import { generateStyle, removeStyle } from "../../utils/importStyles";
import Setting from "./Setting";
import "./Setting.scss";
import path from "path";

test("component renders", () => {
  const wrapper = shallow(<Setting active={true} />);
  expect(wrapper).toMatchSnapshot();
});

test("should recieve props", () => {
  const wrapper = mount(<Setting active={true} />);
  expect(wrapper.prop("active")).toEqual(true);

  wrapper.setProps({active: false});
  expect(wrapper.prop("active")).toEqual(false);

  wrapper.unmount();
});

test("should look proper when checked", async () => {
  const div = document.createElement("div");
  document.body.appendChild(div);
  ReactDOM.render(<Setting active={true} />, div);
  await generateStyle(
    {
      filePaths: [
        path.normalize(__dirname + "/Setting.scss"),
        path.normalize(__dirname + "/../SwitchButton/SwitchButton.scss")
      ],
      // I have no idea why this is working and normal implementations don't
      includePaths: [path.normalize(__dirname + "master.scss")],
    }
  )

  const screenshot = await getComponentImage();
  expect(screenshot).toMatchImageSnapshot();

  // console.log(document.documentElement.outerHTML);

  removeStyle();
  ReactDOM.unmountComponentAtNode(div);
  document.body.removeChild(div);
});

test("should look proper when unchecked", async () => {
  const div = document.createElement("div");
  document.body.appendChild(div);
  ReactDOM.render(<Setting active={false} />, div);
  await generateStyle(
    {
      filePaths: [
        path.normalize(__dirname + "/Setting.scss"),
        path.normalize(__dirname + "/../SwitchButton/SwitchButton.scss")
      ],
      // I have no idea why this is working and normal implementations don't
      includePaths: [path.normalize(__dirname + "master.scss")],
    }
  )

  const screenshot = await getComponentImage();
  expect(screenshot).toMatchImageSnapshot();

  // console.log(document.documentElement.outerHTML);

  removeStyle();
  ReactDOM.unmountComponentAtNode(div);
  document.body.removeChild(div);
});

test("should look proper with description", async () => {
  const div = document.createElement("div");
  document.body.appendChild(div);
  ReactDOM.render(<Setting active={true}>Example description</Setting>, div);
  await generateStyle(
    {
      filePaths: [
        path.normalize(__dirname + "/Setting.scss"),
        path.normalize(__dirname + "/../SwitchButton/SwitchButton.scss")
      ],
      // I have no idea why this is working and normal implementations don't
      includePaths: [path.normalize(__dirname + "master.scss")],
    }
  )

  const screenshot = await getComponentImage();
  expect(screenshot).toMatchImageSnapshot();

  // console.log(document.documentElement.outerHTML);

  removeStyle();
  ReactDOM.unmountComponentAtNode(div);
  document.body.removeChild(div);
});
