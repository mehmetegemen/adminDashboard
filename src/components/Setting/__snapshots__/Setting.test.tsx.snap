// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component renders 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Setting
    active={true}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="setting__content"
        />,
        <div
          className="setting__switch"
        >
          <SwitchButton
            isChecked={true}
          />
        </div>,
      ],
      "className": "setting",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": undefined,
          "className": "setting__content",
        },
        "ref": null,
        "rendered": null,
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <SwitchButton
            isChecked={true}
          />,
          "className": "setting__switch",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "isChecked": true,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="setting__content"
          />,
          <div
            className="setting__switch"
          >
            <SwitchButton
              isChecked={true}
            />
          </div>,
        ],
        "className": "setting",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": undefined,
            "className": "setting__content",
          },
          "ref": null,
          "rendered": null,
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <SwitchButton
              isChecked={true}
            />,
            "className": "setting__switch",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "isChecked": true,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
